File: C++/GOLD/greedy/17420.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int duration[100001];
int willuse[100001];
// pair<int, int> schedule[100001];
vector<pair<int, int>> schedule {};
int main(){
    ios::sync_with_stdio(false), cin.tie(0), cin.tie(0);
    ifstream cin("17420.txt", ios_base::in);    
    int N;
    cin >> N;
    long long ans = 0;
    for (int i = 0; i < N; i++)
    {
        cin >> duration[i];
    }
    for (int i = 0; i < N; i++)
    {
        cin >> willuse[i];
    }
    for (int i = 0; i < N; i++)
    {
        schedule.push_back(make_pair(willuse[i], duration[i]));
        // 이때 사용될 것 // 남은 기한
    }
    sort(schedule.begin(), schedule.end());

    int maxVal = schedule[0].second;
    int maxDuration = schedule[0].first;
    int prevmaxVal = 0;
    int prevmasDuration = 0;
    for (int i = 0; i < N; i++)
    {
        if (maxDuration != schedule[i].first){

            if ( schedule[i].second < maxVal || schedule[i].second < schedule[i].first){

                int tmpMax = max(schedule[i].first, maxVal);    

                
                
                int num = abs(schedule[i].second - tmpMax) / 30 + bool(abs(schedule[i].second - tmpMax) % 30);
                ans += num;

                prevmaxVal = maxVal;
                prevmasDuration = maxDuration;
                maxVal = num * 30 + schedule[i].second;
                maxDuration = schedule[i].first;

            }
            else 
            {
                // 
                maxVal = max(maxVal, schedule[i].second);
            }
        } 
        else 
        {
            if ( schedule[i].second < prevmaxVal || schedule[i].second < schedule[i].first){
                int tmpMax = max(schedule[i].first, prevmaxVal); 
                int num = abs(schedule[i].second - tmpMax) / 30 + bool(abs(schedule[i].second - tmpMax) % 30);
                ans += num;
                maxVal = max(num * 30 + schedule[i].second, maxVal);
            }   
            maxVal = max(maxVal, schedule[i].second);

        }
        // cout << ans << ' ' << maxVal << '\n';
    }
    
    cout << ans;



}

File: Gold/그리디/17420.py
import sys
sys.stdin = open('17420.txt','r')

N = int(input())
duration = list(map(int, input().split()))
will_use = list(map(int, input().split()))

schedule = [(will_use[i], duration[i]) for i in range(N)]
schedule.sort()

ans = 0
max_val = schedule[0][1]
max_duration = schedule[0][0]
prev_max_val = 0
prev_max_duration = 0

for i in range(N):
    if max_duration != schedule[i][0]:
        if schedule[i][1] < max_val or schedule[i][1] < schedule[i][0]:
            tmp_max = max(schedule[i][0], max_val)
            num = abs(schedule[i][1] - tmp_max) // 30 + bool(abs(schedule[i][1] - tmp_max) % 30)
            ans += num

            prev_max_val = max_val
            prev_max_duration = max_duration
            max_val = num * 30 + schedule[i][1]
            max_duration = schedule[i][0]
        else:
            max_val = max(max_val, schedule[i][1])
    else:
        if schedule[i][1] < prev_max_val or schedule[i][1] < schedule[i][0]:
            tmp_max = max(schedule[i][0], prev_max_val)
            num = abs(schedule[i][1] - tmp_max) // 30 + bool(abs(schedule[i][1] - tmp_max) % 30)
            ans += num
            max_val = max(num * 30 + schedule[i][1], max_val)

        max_val = max(max_val, schedule[i][1])

print(ans)

